libsolidity/analysis/TypeChecker.cpp

```diff
@@ -789,27 +788,49 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 
 			solAssert(!dynamic_cast<FixedPointType const*>(var->type()), "FixedPointType not implemented.");
 
-			if (requiresStorage)
+			if (!identifierInfo.suffix.empty())
 			{
-				if (!var->isStateVariable() && !var->type()->dataStoredIn(DataLocation::Storage))
+				string const& suffix = identifierInfo.suffix;
+				solAssert((set<string>{"offset", "slot", "length"}).count(suffix), "");
+				if (var->isStateVariable() || var->type()->dataStoredIn(DataLocation::Storage))
 				{
-					m_errorReporter.typeError(3622_error, _identifier.location, "The suffixes .offset and .slot can only be used on storage variables.");
-					return false;
-				}
-				else if (_context == yul::IdentifierContext::LValue)
-				{
-					if (var->isStateVariable())
+					if (suffix == "length")
 					{
-						m_errorReporter.typeError(4713_error, _identifier.location, "State variables cannot be assigned to - you have to use \"sstore()\".");
+						m_errorReporter.typeError(1856_error, _identifier.location, "State variables only support \".slot\" and \".offset\".");
 						return false;
 					}
-					else if (identifierInfo.isOffset)
+					else if (_context == yul::IdentifierContext::LValue)
 					{
-						m_errorReporter.typeError(9739_error, _identifier.location, "Only .slot can be assigned to.");
+						if (var->isStateVariable())
+						{
+							m_errorReporter.typeError(4713_error, _identifier.location, "State variables cannot be assigned to - you have to use \"sstore()\".");
+							return false;
+						}
+						else if (suffix != "offset")
+						{
+							m_errorReporter.typeError(9739_error, _identifier.location, "Only .slot can be assigned to.");
+							return false;
+						}
+						else
+							solAssert(suffix == "slot", "");
+					}
+				}
+				else if (
+					// bytes calldata or string calldata
+					*var->type() == ArrayType(DataLocation::CallData, true) ||
+					*var->type() == ArrayType(DataLocation::CallData, false)
+				)
+				{
+					if (suffix == "slot")
```

Comment:
I'd rather go with `suffix != "offset" && suffix != "length"` to avoid the problem of adding new suffixes and forgetting to update these checks. Same applies for the above case with `suffix == length`.

Reply:
This one will fire in that case: https://github.com/ethereum/solidity/pull/10208/files#diff-1286fd2f351e6a40b993189fde18c5db99ddfe4a35f4ada894945ba62819a324R794

Reply:
I can add another assertion here if you like.

Reply:
Oh now I see! Yes, will change.